src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs

```diff
@@ -0,0 +1,193 @@
+        /// Enumerate a directory at the given path, if any.
+        /// </summary>
+        /// <param name="subpath">The path that identifies the directory</param>
+        /// <returns>Contents of the directory. Caller must check Exists property.
+        /// The content is a merge of the contents of the provided <see cref="IFileProvider"/>.
+        /// When there is multiple <see cref="IFileInfo"/> with the same Name property, only the first one is included on the results.</returns>
+        public IDirectoryContents GetDirectoryContents(string subpath)
+        {
+            // gets the content of all directories and merged them
+            var existingDirectoryContents = new List<IDirectoryContents>();
+            foreach (var fileProvider in _fileProviders)
+            {
+                var directoryContents = fileProvider.GetDirectoryContents(subpath);
+                if (directoryContents != null && directoryContents.Exists)
+                {
+                    existingDirectoryContents.Add(directoryContents);
+                }
+            }
+
+            // There is no existing directory contents
+            if (existingDirectoryContents.Count == 0)
+            {
+                return new NotFoundDirectoryContents();
+            }
+            var combinedDirectoryContents = new CombinedDirectoryContents(existingDirectoryContents);
+            return combinedDirectoryContents;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="IChangeToken"/> for the specified <paramref name="filter"/>.
+        /// </summary>
+        /// <remarks></remarks>
+        /// <param name="pattern">Filter string used to determine what files or folders to monitor. Example: **/*.cs, *.*, subFolder/**/*.cshtml.</param>
+        /// <returns>An <see cref="IChangeToken"/> that is notified when a file matching <paramref name="filter"/> is added, modified or deleted.
+        /// The change token will be notified when one of the change token returned by the provided <see cref="IFileProvider"/> will be notified.</returns>
+        public IChangeToken Watch(string pattern)
+        {
+            // Watch all file providers
+            var activeChangeTokens = new List<IChangeToken>();
+            foreach (var fileProvider in _fileProviders)
+            {
+                var changeToken = fileProvider.Watch(pattern);
+                if (changeToken != null && changeToken.ActiveChangeCallbacks)
+                {
+                    activeChangeTokens.Add(changeToken);
+                }
+            }
+
+            // There is no change token with active change callbacks
+            if (activeChangeTokens.Count == 0)
+            {
+                return NoopChangeToken.Singleton;
+            }
+            var combinedFileChangeToken = new CombinedFileChangeToken(activeChangeTokens);
+            return combinedFileChangeToken;
+        }
+
+        private class CombinedDirectoryContents : IDirectoryContents
```

Comment:
I'm perhaps alone with this opinion and I've seen this in the other FileSystem projects already, but can't we stop with these nested classes and just make it an internal class within an own file?